src/Slist.hs

```diff
@@ -1872,3 +1880,90 @@ insert = insertBy compare
 insertBy :: (a -> a -> Ordering) -> a -> Slist a -> Slist a
 insertBy f a Slist{..} = Slist (L.insertBy f a sList) (sSize + 1)
 {-# INLINE insertBy #-}
+
+
+----------------------------------------------------------------------------
+-- Generic Functions
+----------------------------------------------------------------------------
+
+{- | @O(1)
+The 'genericLength' function is an overloaded version of 'length'. Instead
+of returning an 'Int', it returns any type which is an instance of 'Num'. 
+In case of an infinite list, 'genericLength' will return -1.
+-}
+genericLength :: Num i => Slist a -> i
+genericLength Slist{..} = case sSize of
+    Infinity -> -1
+    Size n   -> fromIntegral n
+{-# INLINE genericLength #-}
+
+{- | @O(i) | i < n@ and @O(1) | otherwise@.
+The 'genericTake' function is an overloaded version of 'take'. It
+accepts any 'Integral' value as the number of elements to take.
+-}
+genericTake :: Integral i => i -> Slist a -> Slist a
+genericTake i Slist{..}
+    | i <= 0 = mempty
+    | otherwise = Slist
+        { sList = L.genericTake i sList
+        , sSize = Size (fromIntegral i)
```

Comment:
Generally speaking, this is not true, as there are no guarantees that this `i` is always less than the slist itself.